### ISSUE TITLE
Hyperband doesn't use best checkpoints from the previous round

### ISSUE THREAD

**Issue by EvsanDlg** [keras-team/keras-tuner#372]
As far as I could understand from the keras-tuner sources, this realization of HyperBand uses the set of best hyperparameters from the previous rounds, but not that models' weights/checkpoints. Models for new rounds initializes with random weights.
Is this behaviour consistent with the original work? Are there some plans to change this behaviour?

Any way, how could one add corresponding checkpoints' loading at the beginning of rounds to the existing workflow of KerasTuner? Any minimal working example would be appreciated!
Thanks!

**UPD.** Ok, I managed to create my own solution by subclassing Hyperband tuner and overwriting `_on_train_begin`, that fires just before `model.fit` in `run_trial` method:
```
from kerastuner.engine import hypermodel as hm_module

class MyTuner(Hyperband):
    def _on_train_begin(self, model, hp, *fit_args, **fit_kwargs):
        prev_trial_id = hp.values['tuner/trial_id'] if 'tuner/trial_id' in hp else None
        if prev_trial_id:
            prev_trial = self.oracle.trials[prev_trial_id]
            best_epoch = prev_trial.best_step
            # the code below is from load_model method of Tuner class
            with hm_module.maybe_distribute(self.distribution_strategy):
                model.load_weights(self._get_checkpoint_fname(
                    prev_trial.trial_id, best_epoch))
```
Check it out. It seens to work fine, as can be seen at tensorflow acc plot:
![image](https://user-images.githubusercontent.com/15376599/89208669-7798d600-d5c5-11ea-9d2f-5da5d1b33568.png)
Any comments are appreciated

**Comment by yixingfu** [comment]
This looks very good!

But I don't think the original [paper for hyperband](https://arxiv.org/pdf/1603.06560.pdf) suggest specifically about whether to use weight from previous round (or maybe I just went through the paper too carelessly)? Or is there any newer work suggesting this approach is better that you can point me to?

In general it may be interesting to include an option to use previous training result for further trials, but it probably not always good (sometimes early trial result may mislead).

**Comment by EvsanDlg** [comment]
@yixingfu , sorry that my answer took so long. I saw the mention of such approach in [Kevin Jamieson's article at Washington University website](https://homes.cs.washington.edu/~jamieson/hyperband.html). It's written that:
_HYPERBAND requires 1) the ability to sample a hyperparameter configuration (get_random_hyperparameter_configuration()), and 2) the ability to train a particular hyperparameter configuration until it has reached a given number of iterations **(perhaps resuming from a previous checkpoint)**..
Anyway, it seems to great if keras tuner have a built-in functionality to choose this behaviour

**Comment by TsaiTung-Chen** [comment]
I'm facing the same issue too.
Here's my metric history on the training set (10 colors for 10 hyper-configs):
![SimpleCNN_bracket 2 (training)](https://user-images.githubusercontent.com/62462368/112014003-796bd580-8b65-11eb-8648-5b3e740b562c.png)

As you can see, every hyper-config model seems to restart training when entering a new round.
(max_epochs (R) = 3^3, factor (eta) = 3, with early stopping)
And the dataset and model are from [here](https://www.tensorflow.org/tutorials/keras/keras_tuner).

**Comment by otkernell** [comment]
Agree with @EvsanDlg. A major plus of hyperband is that it reduces the number of iterations. While it might not be called out directly in the paper, the experiments seem to imply it.

**Comment by TimMensch** [comment]
For posterity:

1. The above solution doesn't work because the Tuner class changed.
2. `Hyperband` was at some point updated to perform this reloading, but its code wasn't functional either because *the Tuner class changed again.*
3. For the current version of `Hyperband`, this seems to work:

```python
class HyperbandPlus(Hyperband):
    def _build_hypermodel(self, hp):
        model = super(Hyperband, self)._build_hypermodel(hp)
        if "tuner/trial_id" in hp.values:
            trial_id = hp.values["tuner/trial_id"]
            print("Reloading data from", trial_id)
            fname = self._get_checkpoint_fname(trial_id)
            # Load best checkpoint from this trial.
            model.load_weights(fname)
        return model
```

This is the code from the current `Hyperband` implementation, updated to override `_build_hypermodel` instead of `_build_model`, which no longer exists. Aside: This is why static types are important. If this code had consistent type definitions and had a type linting build step, as soon as `_build_model` was renamed the `Hyperband` code would have thrown an error and wouldn't have just silently broken.

I think the `self.oracle.get_trial(trial_id)` is redundant, since the `trial_id` in the oracle appears to be the same as the `trial_id`, but haven't cleaned it up.

When I get a chance I'll submit the above fix as a PR, but until then I wanted to post the fix in case anyone else was wondering why their `Hyperband` trials weren't reloading weights.

**Comment by TylerADavis** [comment]
Is this issue resolved now that https://github.com/keras-team/keras-tuner/pull/711 has been merged?

**Comment by haifeng-jin** [comment]
@TylerADavis Yes, but it has not been released yet. You may use master branch for now. Thanks.

Closing the issue.
